contributors/design-proposals/core-metrics-pipeline.md

```diff
@@ -0,0 +1,295 @@
+  // Stats pertaining to volume usage of filesystem resources.  
+  // +optional  
+  Volumes []VolumeUsage `json:"volume,omitempty" patchStrategy:"merge" patchMergeKey:"name"`  
+}  
+
+// ContainerUsage holds container-level usage stats.  
+type ContainerUsage struct {  
+  // UID of the container  
+  ContainerUID string `json:"uid"`  
+  // Stats pertaining to container usage of cpu  
+  // +optional  
+  CPU *CpuUsage `json:"memory,omitempty"`  
+  // Stats pertaining to container usage of system memory  
+  // +optional  
+  Memory *MemoryUsage `json:"memory,omitempty"`  
+  // Stats pertaining to container rootfs usage of disk.  
+  // Rootfs.UsedBytes is the number of bytes used for the container write layer.  
+  // +optional  
+  Rootfs *FilesystemUsage `json:"rootfs,omitempty"`  
+  // Stats pertaining to container logs usage of Disk.  
+  // +optional  
+  Logs *FilesystemUsage `json:"logs,omitempty"`  
+}  
+
+// CpuUsage holds statistics about the amount of cpu time consumed  
+type CpuUsage struct {  
+  // The time at which these stats were updated.  
+  Timestamp metav1.Time `json:"time"`  
+  // Average CPU usage rate over sample window (across all cores), in "cores".  
+  // The "core" unit represents nanoseconds of CPU time consumed per second.  
+  // For example, 5 nanocores means the process averaged 5 nanoseconds 
+  // of cpu time per second during the sample window.
+  // +optional  
+  UsageRateNanoCores *uint64 `json:"usageNanoCores,omitempty"`  
+  // Cumulative CPU usage (sum of all cores) since object creation.  
+  // +optional  
+  AggregateUsageCoreNanoSeconds *uint64 `json:"usageCoreNanoSeconds,omitempty"`  
+}  
+
+// MemoryUsage holds statistics about the quantity of memory consumed  
+type MemoryUsage struct {  
+  // The time at which these stats were updated.  
+  Timestamp metav1.Time `json:"time"`  
+  // The amount of working set memory. This includes recently accessed memory,  
+  // dirty memory, and kernel memory.  
+  // +optional  
+  UsageBytes *uint64 `json:"usageBytes,omitempty"`  
+}  
+
+// VolumeUsage holds statistics about the quantity of disk resources consumed for a volume  
+type VolumeUsage struct {  
+  // Embedded FilesystemUsage  
+  FilesystemUsage  
+  // Name is the name given to the Volume  
+  // +optional  
+  Name string `json:"name,omitempty"`  
+}  
+
+// FilesystemUsage holds statistics about the quantity of disk resources consumed  
+type FilesystemUsage struct {  
+  // The time at which these stats were updated.  
+  Timestamp metav1.Time `json:"time"`  
+  // The device on which resources are consumed  
+  Device string `json:"device"`  
+  // UsedBytes represents the disk space consumed on the device, in bytes.  
+  // +optional  
+  UsedBytes *uint64 `json:"usedBytes,omitempty"`  
+  // InodesUsed represents the inodes consumed on the device  
+  // +optional  
+  InodesUsed *uint64 `json:"inodesUsed,omitempty"`  
+}  
+```
```

Comment:
You talked about gathering stats on demand, but some use cases (e.g. kubectl top) also require regular updates. I think it would be worth adding a section about how this will work. For example, every stat request could include a "required freshness", and pulled from cache if there are recent stats, or fetch new ones if not.